ddl/table.go

```diff
@@ -948,56 +956,189 @@ func updateVersionAndTableInfo(t *meta.Meta, job *model.Job, tblInfo *model.Tabl
 	return ver, t.UpdateTable(job.SchemaID, tblInfo)
 }
 
-// TODO: It may have the issue when two clients concurrently add partitions to a table.
-func onAddTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+func checkAddPartition(t *meta.Meta, job *model.Job) (*model.TableInfo, *model.PartitionInfo, []*model.PartitionDefinition, error) {
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, nil, errors.Trace(err)
+	}
 	partInfo := &model.PartitionInfo{}
-	err := job.DecodeArgs(&partInfo)
+	err = job.DecodeArgs(&partInfo)
 	if err != nil {
 		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+		return nil, nil, nil, errors.Trace(err)
+	}
+	partDefPointers := make([]*model.PartitionDefinition, 0, len(partInfo.Definitions))
+	for _, pd := range partInfo.Definitions {
+		partitionDefInMeta := tblInfo.FindPartitionDefinitionByName(pd.Name.L)
+		if partitionDefInMeta != nil {
+			if partitionDefInMeta.State == model.StatePublic {
+				// Even one partition with the same name existed, whole job canceled.
+				// We already have these new added partition with the same name.
+				job.State = model.JobStateCancelled
+				return nil, nil, nil, errors.Trace(ErrSameNamePartition.GenWithStackByArgs(pd.Name.L))
+			}
+			// Collect the partition definition pointer in tableInfo for future state change.
+			partDefPointers = append(partDefPointers, partitionDefInMeta)
+		}
 	}
+	return tblInfo, partInfo, partDefPointers, nil
+}
 
-	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+func checkPartitionReplica(partDefPointers []*model.PartitionDefinition, d *ddlCtx, job *model.Job) (needWait bool, err error) {
+	ctx := context.Background()
+	pdCli := d.store.(tikv.Storage).GetRegionCache().PDClient()
+	stores, err := pdCli.GetAllStores(ctx)
 	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+		return needWait, errors.Trace(err)
 	}
-	err = checkAddPartitionTooManyPartitions(uint64(len(tblInfo.Partition.Definitions) + len(partInfo.Definitions)))
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+	for _, pd := range partDefPointers {
+		startKey, endKey := tablecodec.GetTableHandleKeyRange(pd.ID)
+		regions, _, err := pdCli.ScanRegions(ctx, startKey, endKey, -1)
+		if err != nil {
+			return needWait, errors.Trace(err)
+		}
+		// For every region in the partition, if it has some corresponding peers and
+		// no pending peers, that means the replication has completed.
```

Comment:
Why not query it from tableinfo?

Reply:
fiflash state machine is same to tikv state machine, they will report to PD